# GROUP MEMBERS: Mario Ruiz Vaquett, Javier Pérez Vargas

## Libraries

In [1]:
import pandas as pd

In [2]:
data1 = pd.read_excel("hospital1.xlsx")
data2 = pd.read_excel("hospital2.xlsx")

# Preprocessing data 1 (hospital1.xlsx)

In [3]:
data1.shape

(14712, 54)

In [4]:
data1.head()

patient ID  patient ID.1 nationality  age gender K=female E=male  \
0    11850006      11850006        T.C.   47                      K   
1     4373630       4373630        T.C.   23                      K   
2    12172023      12172023        T.C.   26                      K   
3    72856168      72856168        T.C.   60                      K   
4    72200559      72200559        T.C.   72                      K   

  date_of_first_symptoms BASVURUTARIHI  fever_temperature  oxygen_saturation  \
0             2022-01-01    2022-01-01               38.2               96.0   
1             2022-01-01    2022-01-01               37.0               96.0   
2             2022-01-01    2022-01-01               38.0               95.0   
3             2022-01-01    2022-01-01               36.5               95.0   
4             2021-12-22    2021-12-22               36.9               96.0   

   history_of_fever  ...  chronic_hematologic_disease  AIDS_HIV  \
0                 0  ...                          0.0       0.0   
1                 1  ...                          0.0       0.0   
2                 1  ...                          0.0       0.0   
3                 0  ...                          0.0       0.0   
4                 0  ...                          0.0       0.0   

   diabetes_mellitus_type_1  diabetes_mellitus_type_2  rheumatologic_disorder  \
0                       0.0                       0.0                     0.0   
1                       0.0                       0.0                     0.0   
2                       0.0                       0.0                     0.0   
3                       0.0                       0.0                     0.0   
4                       0.0                       0.0                     0.0   

   dementia  tuberculosis  smoking  other_risks  PCR_result  
0       0.0             0        0            0    positive  
1       0.0             0        0            0    positive  
2       0.0             0        0            0    positive  
3       0.0             0        0            0    positive  
4       0.0             0        0            0    positive  

[5 rows x 54 columns]

## ¿Columnas redundantes?

En primer lugar, vemos que existe una columna con un nombre extravagante, cuyos valores parecen iguales a los de la columna que define la fecha de los primeros síntomas "date_of_first_symptoms"

Su nombre es BASVURUTAHIRI, y con un traductor detectamos que está en idioma turco y que significa "Fecha de solicitud" en castellano. No es exactamente lo mismo que "date_of_first_symptoms", así que veamos si difieren ena alguna fila.

In [5]:
data1.loc[(data1['date_of_first_symptoms'] != data1['BASVURUTARIHI'])]

Empty DataFrame
Columns: [patient ID, patient ID.1, nationality, age, gender K=female E=male, date_of_first_symptoms, BASVURUTARIHI, fever_temperature, oxygen_saturation, history_of_fever, cough, sore_throat, runny_nose, wheezing, shortness_of_breath, lower_chest_wall_indrawing, chest_pain, conjunctivitis, lymphadenopathy, headache, loss_of_smell, loss_of_taste, fatigue_malaise, anorexia, altered_consciousness_confusion, muscle_aches, joint_pain, inability_to_walk, abdominal_pain, diarrhoea, vomiting_nausea, skin_rash, bleeding, other_symptoms, chronic_cardiac_disease, hypertension, chronic_pulmonary_disease, asthma, chronic_kidney_disease, obesity, liver_disease, asplenia, chronic_neurological_disorder, malignant_neoplasm, chronic_hematologic_disease, AIDS_HIV, diabetes_mellitus_type_1, diabetes_mellitus_type_2, rheumatologic_disorder, dementia, tuberculosis, smoking, other_risks, PCR_result]
Index: []

[0 rows x 54 columns]

Como vemos, no hay ninguna columna donde haya valores diferentes, por tanto, podemos decir con certeza que la columna BASVURUTARIHI es igual a la anterior. Por tanto, procedemos a eliminarla.

In [6]:
data1.drop(['BASVURUTARIHI'], axis=1, inplace=True)
data1.head(3)

patient ID  patient ID.1 nationality  age gender K=female E=male  \
0    11850006      11850006        T.C.   47                      K   
1     4373630       4373630        T.C.   23                      K   
2    12172023      12172023        T.C.   26                      K   

  date_of_first_symptoms  fever_temperature  oxygen_saturation  \
0             2022-01-01               38.2               96.0   
1             2022-01-01               37.0               96.0   
2             2022-01-01               38.0               95.0   

   history_of_fever  cough  ...  chronic_hematologic_disease  AIDS_HIV  \
0                 0      0  ...                          0.0       0.0   
1                 1      0  ...                          0.0       0.0   
2                 1      1  ...                          0.0       0.0   

   diabetes_mellitus_type_1  diabetes_mellitus_type_2  rheumatologic_disorder  \
0                       0.0                       0.0                     0.0   
1                       0.0                       0.0                     0.0   
2                       0.0                       0.0                     0.0   

   dementia  tuberculosis  smoking  other_risks  PCR_result  
0       0.0             0        0            0    positive  
1       0.0             0        0            0    positive  
2       0.0             0        0            0    positive  

[3 rows x 53 columns]

En segundo lugar, vemos que hay 2 columnas asociadas al ID del paciente: "patient ID" y "patient ID.1". De la misma manera que antes, veamos si difieren en alguna fila.

In [7]:
data1.loc[data1["patient ID"] != data1["patient ID.1"]]

patient ID  patient ID.1 nationality  age gender K=female E=male  \
1187       712249       6798097        T.C.   28                      K   
2957       215634      98284721        T.C.   33                      E   
4219      2015016      96844453        T.C.   51                      E   
6470       712249       6798097        T.C.   30                      K   
7774      2015016      96844453        T.C.   52                      E   
11900      712249       6798097        T.C.   28                      K   
13616      712249       6798097        T.C.   28                      K   

      date_of_first_symptoms  fever_temperature  oxygen_saturation  \
1187              2022-02-03               38.6               95.0   
2957              2021-05-06                NaN               86.0   
4219              2021-09-08               39.2               95.0   
6470              2022-01-13               36.8               95.0   
7774              2021-10-15               37.8               95.0   
11900             2021-09-25               38.3               96.0   
13616             2021-06-29               36.5               91.0   

       history_of_fever  cough  ...  chronic_hematologic_disease  AIDS_HIV  \
1187                  1      1  ...                          0.0       0.0   
2957                  0      0  ...                          0.0       0.0   
4219                  1      1  ...                          0.0       0.0   
6470                  0      1  ...                          0.0       0.0   
7774                  0      0  ...                          0.0       0.0   
11900                 0      0  ...                          0.0       0.0   
13616                 1      1  ...                          0.0       0.0   

       diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
1187                        0.0                       0.0   
2957                        0.0                       0.0   
4219                        0.0                       0.0   
6470                        0.0                       0.0   
7774                        0.0                       0.0   
11900                       0.0                       0.0   
13616                       0.0                       0.0   

       rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
1187                      0.0       0.0             0        0            0   
2957                      0.0       0.0             0        0            0   
4219                      0.0       0.0             0        0            0   
6470                      0.0       0.0             0        0            0   
7774                      0.0       0.0             0        0            0   
11900                     0.0       0.0             0        0            0   
13616                     0.0       0.0             0        0            0   

       PCR_result  
1187     positive  
2957     positive  
4219          NaN  
6470     positive  
7774     positive  
11900    positive  
13616         NaN  

[7 rows x 53 columns]

En este caso sí, hay 7 filas en los que las columnas difieren. En principio, a pesar de que tengan diferentes IDs, la combinación de ambos IDs podría actuar como un ID único, pues por nacionalidad, edad y género podrían ser la misma persona. Es decir, la combinación (712249, 6798097), que aparece 4 veces, parece ser la misma persona, aunque la razón de que tenga diferentes IDs es desconocida. De la misma forma, la combinación (2015016, 96844453) parecen ser la misma persona. De (215634, 98284721) solo tenemos un registro.

Ante esta situación, debemos plantear 4 posibles situaciones que darían problemas:
- P1: De entre los IDs de 'patient ID' que difieren, ¿son IDs que se corresponden con alguna otra persona en 'patient ID.1'?
- P2: De entre los IDs de 'patient ID.1' que difieren, ¿son IDs que se corresponden con alguna otra persona en 'patient ID'?

***Observación: No tiene sentido hacerse la pregunta "De entre los IDs de 'patient ID' que difieren, ¿son IDs que se corresponden con alguna otra persona en 'patient ID'?", pues, de ser la respuesta afirmativa, habría dos casos: o bien esa otra persona tiene IDs distintos y se habría identificado arriba, o bien tienen los mismos IDs y se identificará en P2. De la misma manera se podría justificar con 'patient ID.1'.

Comprobemos cada caso

In [8]:
different_IDs_patientID = data1.loc[data1["patient ID"] != data1["patient ID.1"]]['patient ID']
different_IDs_patientID1 = data1.loc[data1["patient ID"] != data1["patient ID.1"]]['patient ID.1']

#### P1

In [9]:
data1.loc[data1['patient ID'].isin(different_IDs_patientID1)]

patient ID  patient ID.1 nationality  age gender K=female E=male  \
8378     98284721      98284721        T.C.   69                      E   
14440     6798097       6798097        T.C.   59                      K   

      date_of_first_symptoms  fever_temperature  oxygen_saturation  \
8378              2022-01-17               36.5               89.0   
14440             2021-07-31               36.7               97.0   

       history_of_fever  cough  ...  chronic_hematologic_disease  AIDS_HIV  \
8378                  0      0  ...                          0.0       0.0   
14440                 0      0  ...                          0.0       0.0   

       diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
8378                        0.0                       0.0   
14440                       0.0                       0.0   

       rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
8378                      0.0       0.0             0        0            0   
14440                     0.0       0.0             0        0            0   

       PCR_result  
8378     positive  
14440    positive  

[2 rows x 53 columns]

#### P2

In [10]:
data1.loc[data1['patient ID.1'].isin(different_IDs_patientID)]

Empty DataFrame
Columns: [patient ID, patient ID.1, nationality, age, gender K=female E=male, date_of_first_symptoms, fever_temperature, oxygen_saturation, history_of_fever, cough, sore_throat, runny_nose, wheezing, shortness_of_breath, lower_chest_wall_indrawing, chest_pain, conjunctivitis, lymphadenopathy, headache, loss_of_smell, loss_of_taste, fatigue_malaise, anorexia, altered_consciousness_confusion, muscle_aches, joint_pain, inability_to_walk, abdominal_pain, diarrhoea, vomiting_nausea, skin_rash, bleeding, other_symptoms, chronic_cardiac_disease, hypertension, chronic_pulmonary_disease, asthma, chronic_kidney_disease, obesity, liver_disease, asplenia, chronic_neurological_disorder, malignant_neoplasm, chronic_hematologic_disease, AIDS_HIV, diabetes_mellitus_type_1, diabetes_mellitus_type_2, rheumatologic_disorder, dementia, tuberculosis, smoking, other_risks, PCR_result]
Index: []

[0 rows x 53 columns]

Efectivamente, observamos que hay dos filas problemáticas reveladas en P1. Son las filas que tienen los siguiente IDs: 98284721 y 6798097. Por tanto, tenemos el siguiente problema en estos dos casos:
- Existe 1 fila con ID 98284721 en ambas columnas, y existen 1 fila con ID 215634 en 'patient ID' y con ID 98284721 en 'patient ID.1'. El primero de ellos es turco, tiene 69 años y es hombre, mientras que el otro registro dice que es un hombre turco de 33 años. Asumiendo que el dato de edad se recogió correctamente, podemos asegurar que son personas diferentes.
  
- Existe 1 fila con ID 6798097 en ambas columnas, y existen 4 filas con ID 712249 en 'patient ID' y con ID 6798097 en 'patient ID.1'. El primero es una mujer turca de 59 años, mientras que el segundo es una mujer turca de entre 28 y 30 años (*). De nuevo, con la misma asunción que antes, podemos asegurar que son diferentes personas.

Vemos que, a pesar de compartir ID en 'patient ID.1', son diferentes personas. Esto implica que la columna que introduce incoherencias en los datos es 'patient ID.1'. Al contrario, en la columna 'patient ID' todas las personas distintas tienen diferente ID. Por tanto, la solución será eliminar la columna 'patient ID.1'.

\* Viendo los datos hemos observado, además, incoherencias en la edad de personas que son la misma. Posteriormente trataremos este problema. 

In [11]:
data1.drop(['patient ID.1'], axis=1, inplace=True)
data1.head(3)

patient ID nationality  age gender K=female E=male date_of_first_symptoms  \
0    11850006        T.C.   47                      K             2022-01-01   
1     4373630        T.C.   23                      K             2022-01-01   
2    12172023        T.C.   26                      K             2022-01-01   

   fever_temperature  oxygen_saturation  history_of_fever  cough  sore_throat  \
0               38.2               96.0                 0      0            0   
1               37.0               96.0                 1      0            1   
2               38.0               95.0                 1      1            0   

   ...  chronic_hematologic_disease  AIDS_HIV  diabetes_mellitus_type_1  \
0  ...                          0.0       0.0                       0.0   
1  ...                          0.0       0.0                       0.0   
2  ...                          0.0       0.0                       0.0   

   diabetes_mellitus_type_2  rheumatologic_disorder  dementia  tuberculosis  \
0                       0.0                     0.0       0.0             0   
1                       0.0                     0.0       0.0             0   
2                       0.0                     0.0       0.0             0   

   smoking  other_risks  PCR_result  
0        0            0    positive  
1        0            0    positive  
2        0            0    positive  

[3 rows x 52 columns]

### ¿Incoherencias en la edad?

Vamos a tratar el tema de la edad. Por ejemplo, remitiéndonos a uno de las personas anteriores:

In [12]:
data1.loc[data1['patient ID'] == 712249].sort_values(by='date_of_first_symptoms')

patient ID nationality  age gender K=female E=male  \
13616      712249        T.C.   28                      K   
11900      712249        T.C.   28                      K   
6470       712249        T.C.   30                      K   
1187       712249        T.C.   28                      K   

      date_of_first_symptoms  fever_temperature  oxygen_saturation  \
13616             2021-06-29               36.5               91.0   
11900             2021-09-25               38.3               96.0   
6470              2022-01-13               36.8               95.0   
1187              2022-02-03               38.6               95.0   

       history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
13616                 1      1            1  ...                          0.0   
11900                 0      0            0  ...                          0.0   
6470                  0      1            1  ...                          0.0   
1187                  1      1            0  ...                          0.0   

       AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
13616       0.0                       0.0                       0.0   
11900       0.0                       0.0                       0.0   
6470        0.0                       0.0                       0.0   
1187        0.0                       0.0                       0.0   

       rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
13616                     0.0       0.0             0        0            0   
11900                     0.0       0.0             0        0            0   
6470                      0.0       0.0             0        0            0   
1187                      0.0       0.0             0        0            0   

       PCR_result  
13616         NaN  
11900    positive  
6470     positive  
1187     positive  

[4 rows x 52 columns]

Como podemos observar, estas 4 filas parecen referirse a la misma persona, por su coincidencia en el ID, y que son mujeres turcas. No obstante, la edad que aparece no tiene sentido. En menos de 4 meses, pasa de 28 a 30 años, y luego vuelve a tener 28. Esto es algo que hemos observado en más personas, así que veamos que podemos hacer para tratarlo.

En primer lugar, nos interesa ver el rango de las fechas de primeros síntomas.

In [13]:
data1.sort_values(by='date_of_first_symptoms').iloc[[0, -1]]

patient ID nationality  age gender K=female E=male  \
194     18839181        T.C.   48                      K   
7148    87532216        T.C.   77                      K   

     date_of_first_symptoms  fever_temperature  oxygen_saturation  \
194              2021-05-01               37.7               99.0   
7148             2022-03-14               37.6               96.0   

      history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
194                  0      0            0  ...                          0.0   
7148                 1      0            0  ...                          0.0   

      AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
194        0.0                       0.0                       0.0   
7148       0.0                       0.0                       0.0   

      rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
194                      0.0       0.0             0        0            0   
7148                     0.0       0.0             0        0            0   

      PCR_result  
194     positive  
7148    positive  

[2 rows x 52 columns]

Observamos que el primer registro es del 1 de mayo de 2021, mientras que el último es del 14 de marzo de 2022. Es decir, un rango aproximado de 10 meses y medio. En este periodo, las personas que aparezcan más de una vez en el dataset solo han podido cumplir un año de edad. El objetivo será unificar la información de los registros que se refieran a la misma persona, con la posibilidad de que posteriormente se cree una columna relacionada con el número de PCR realizadas que pueda aportar información.

Veamos las diferencias de edad. Para analizar este problema, en primer lugar nos gustaría ver cuáles son las mayores diferencias de edad entre filas con el mismo ID, es decir, en la misma persona. Para ello crearemos una función.

In [14]:
def age_incoherences(min_difference):
    age_diffs = data1.groupby('patient ID')['age'].agg(lambda x: x.max() - x.min())
    count = (age_diffs >= min_difference).sum()
    indexes = age_diffs[age_diffs >= min_difference].index
    return count, indexes

In [15]:
print("Diferencias de 1 o más años: ", age_incoherences(1)[0])
print("Diferencias de 2 o más años: ", age_incoherences(2)[0])
print("Diferencias de 3 o más años: ", age_incoherences(3)[0])

Diferencias de 1 o más años:  1606
Diferencias de 2 o más años:  712
Diferencias de 3 o más años:  0


Así, verificamos que existen 712 personas cuya edad ha sido, con toda seguridad, codificada erróneamente pues, como hemos dicho antes, solo pasan 10 meses y medio desde el primero hasta el último dato. Afortunadamente, esta es la máxima diferencia de edad en todo el dataset, ya que al comprobarlo con 'min_difference' = 3 obtenemos 0 diferencias. Además, hay 894 (1606 - 712) personas que tienen una diferencia de un año entre varios de sus registros, aunque podría ser porque han cumplido años en ese período.

Ante esta situación, debemos modificar las edades de alguna manera para evitar las incoherencias en el dataset. Se nos ocurren dos soluciones:
- Solución A: La solución más básica sería poner la misma edad a todos los personas que repitan PCR. Esta podría ser la más baja, la más alta, la que esté entre las dos (si la diferencia es de 2), o la más repetida. Esta solución evitaría incoherencias de edad.
- Solución B: Una solución un poco más elaborada sería jugar con la edad y con la fecha de primeros síntomas. Es posible (de hecho, toda seguridad, debido a la dimensión del dataset), que existan personas que SÍ hayan cumplido años en esos 10 meses y medio. Para estas personas, se podría(n) mantener la(s) primera(s) fila(s) que tienen la edad más baja, y a partir de la fila en la que la edad está actualizada (asumiendo que ha cumplido un año), actualizar las registros posteriores con la edad actualizada.

La solución B podría ser más correcta, pero nada nos asegura que las edades de personas que, supuestamente, han cumplido años, también se hayan codificado erróneamente. Además, habría que valorar una casuística considerablemente más alta que con la solución A y creemos que una diferencia de un año no va a afectar apenas al modelo final de predicción. Por tanto, asumiremos este "error" y aplicaremos la solución A.

Para aplicar la solución A, nos ayudaremos de la función creada anteriormente y modificaremos la edad de aquellos registros que representen a la misma persona y tengan edades diferentes. Concretamente, elegiremos la edad más repetida en sus registros.

In [16]:
ids_of_people_with_age_incoherences = age_incoherences(1)[1]

Vamos a utilizar la siguiente persona de ejemplo (ID = 292647) para ver si se modifican correctamente las edades.

In [17]:
data1.loc[data1['patient ID'] == ids_of_people_with_age_incoherences[3]]

patient ID nationality  age gender K=female E=male  \
1342      269029        T.C.   40                      K   
8173      269029        T.C.   39                      K   
9028      269029        T.C.   41                      K   

     date_of_first_symptoms  fever_temperature  oxygen_saturation  \
1342             2021-05-03               38.3               97.0   
8173             2021-08-16               38.1               97.0   
9028             2021-08-18               36.6               98.0   

      history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
1342                 1      1            1  ...                          0.0   
8173                 1      1            1  ...                          0.0   
9028                 0      0            0  ...                          0.0   

      AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
1342       0.0                       0.0                       0.0   
8173       0.0                       0.0                       0.0   
9028       0.0                       0.0                       0.0   

      rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
1342                     0.0       0.0             0        0            0   
8173                     0.0       0.0             0        0            0   
9028                     0.0       0.0             0        0            0   

      PCR_result  
1342    positive  
8173    positive  
9028    positive  

[3 rows x 52 columns]

Haciendo un agrupamiento por 'patient ID' y deshaciéndolo con la función transform, podemos obtener la moda de la edad de cada persona y lo aplicamos de la siguiente manera:

In [18]:
def mode_or_min(series):
    return series.mode()[0]

data1['age'] = data1.groupby('patient ID')['age'].transform(mode_or_min)

Comprobamos si se han modificado los valores.

In [19]:
data1.loc[data1['patient ID'] == ids_of_people_with_age_incoherences[3]]

patient ID nationality  age gender K=female E=male  \
1342      269029        T.C.   39                      K   
8173      269029        T.C.   39                      K   
9028      269029        T.C.   39                      K   

     date_of_first_symptoms  fever_temperature  oxygen_saturation  \
1342             2021-05-03               38.3               97.0   
8173             2021-08-16               38.1               97.0   
9028             2021-08-18               36.6               98.0   

      history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
1342                 1      1            1  ...                          0.0   
8173                 1      1            1  ...                          0.0   
9028                 0      0            0  ...                          0.0   

      AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
1342       0.0                       0.0                       0.0   
8173       0.0                       0.0                       0.0   
9028       0.0                       0.0                       0.0   

      rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
1342                     0.0       0.0             0        0            0   
8173                     0.0       0.0             0        0            0   
9028                     0.0       0.0             0        0            0   

      PCR_result  
1342    positive  
8173    positive  
9028    positive  

[3 rows x 52 columns]

Así, podemos observar que los datos se han modificado correctamente y se ha puesto la moda, que eran 26 años. Observemos que 27 también se repetía 3 veces, pero la función coge por defecto la mínima moda. Además, si no hay moda, coge el mínimo de la serie, lo cual nos ha parecido correcto.

Ahora que ya hemos arreglado la edad, nos podríamos plantear la siguiente pregunta: ¿Hay incoherencias en la nacionalidad o en el género?

### Incoherencias en la nacionalidad

Comprobemos si hay registros con el mismo ID pero distinta nacionalidad.

In [20]:
data1.groupby('patient ID')['nationality'].nunique().value_counts()

nationality
1    11093
2       29
Name: count, dtype: int64

De nuevo, hay 29 IDs que contienen alguna incoherencia en la nacionalidad.

In [21]:
ids_of_nationality_incoherences = data1.groupby('patient ID')['nationality'].nunique().loc[lambda x: x > 1].index

In [22]:
data1.loc[data1['patient ID'].isin(ids_of_nationality_incoherences)].sort_values(by='patient ID')

patient ID           nationality  age gender K=female E=male  \
9960      3489578                  T.C.   68                      K   
3880      3489578  United Arab Emirates   68                      K   
10954     3919600             Venezuela   71                      K   
928       3919600                  T.C.   71                      K   
3653      5300371                  T.C.   24                      K   
...           ...                   ...  ...                    ...   
10682    98668372                  T.C.   30                      K   
4975     98668372                  T.C.   30                      K   
10794    98668372           Afghanistan   30                      K   
2153     98668372                  T.C.   30                      K   
4277     98668372                  T.C.   30                      K   

      date_of_first_symptoms  fever_temperature  oxygen_saturation  \
9960              2022-01-21               37.9              100.0   
3880              2022-01-08               36.6               91.0   
10954             2021-06-21               36.9               92.0   
928               2021-10-02               36.8               90.0   
3653              2021-07-07               37.3               96.0   
...                      ...                ...                ...   
10682             2021-10-22               37.2               89.0   
4975              2022-01-10               38.0               90.0   
10794             2021-12-22               36.5               96.0   
2153              2021-11-04               36.8               97.0   
4277              2021-10-08               39.1               94.0   

       history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
9960                  0      0            0  ...                          0.0   
3880                  0      0            0  ...                          0.0   
10954                 0      1            0  ...                          0.0   
928                   0      0            0  ...                          0.0   
3653                  1      1            1  ...                          0.0   
...                 ...    ...          ...  ...                          ...   
10682                 0      0            0  ...                          0.0   
4975                  0      0            0  ...                          0.0   
10794                 0      0            0  ...                          0.0   
2153                  0      0            0  ...                          0.0   
4277                  0      0            0  ...                          0.0   

       AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
9960        0.0                       0.0                       0.0   
3880        0.0                       0.0                       0.0   
10954       0.0                       0.0                       0.0   
928         0.0                       0.0                       0.0   
3653        0.0                       0.0                       0.0   
...         ...                       ...                       ...   
10682       0.0                       0.0                       0.0   
4975        0.0                       0.0                       0.0   
10794       0.0                       0.0                       0.0   
2153        0.0                       0.0                       0.0   
4277        0.0                       0.0                       0.0   

       rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
9960                      0.0       0.0             0        0            0   
3880                      0.0       0.0             0        0            0   
10954                     0.0       0.0             0        0            0   
928                       0.0       0.0             0        0            0   
3653                      0.0       0.0             0        0            0   
...          

Habiendo observado los 29 casos, en absolutamente todos se repite el mismo patrón: un registro de la persona con nacionalidad que no es T.C., y después uno o más registros con la nacionalidad T.C. 

Es posible que la misma persona tenga varias nacionalidades. Lo que sí parece extraño es que una misma persona presente distintas nacionalidades al realizarse una PCR, en vez de presentar siempre la misma. Dado que el hospital parece turco, y T.C. se refiere a la nacionalidad turca, que es la más común, vamos a asumir que estas 29 personas con incoherencias en la nacionalidad son de verdad turcas, y cambiaremos los registros donde se indique lo contrario.

Dicho esto, es cierto que esta simplificación se basa en una asunción que podría ser errónea. No obstante, tras haberlo valorado, hemos pensado que la nacionalidad posiblemente no será uno de los factores más relevantes en la predicción, y que esta simplificación permite evitar posibles incoherencias en el futuro. 

In [23]:
for id in ids_of_nationality_incoherences:
    data1.loc[data1['patient ID'] == id, 'nationality'] = "T.C."

In [24]:
data1.loc[data1['patient ID'].isin(ids_of_nationality_incoherences)].sort_values(by='patient ID')

patient ID nationality  age gender K=female E=male  \
9960      3489578        T.C.   68                      K   
3880      3489578        T.C.   68                      K   
10954     3919600        T.C.   71                      K   
928       3919600        T.C.   71                      K   
3653      5300371        T.C.   24                      K   
...           ...         ...  ...                    ...   
10682    98668372        T.C.   30                      K   
4975     98668372        T.C.   30                      K   
10794    98668372        T.C.   30                      K   
2153     98668372        T.C.   30                      K   
4277     98668372        T.C.   30                      K   

      date_of_first_symptoms  fever_temperature  oxygen_saturation  \
9960              2022-01-21               37.9              100.0   
3880              2022-01-08               36.6               91.0   
10954             2021-06-21               36.9               92.0   
928               2021-10-02               36.8               90.0   
3653              2021-07-07               37.3               96.0   
...                      ...                ...                ...   
10682             2021-10-22               37.2               89.0   
4975              2022-01-10               38.0               90.0   
10794             2021-12-22               36.5               96.0   
2153              2021-11-04               36.8               97.0   
4277              2021-10-08               39.1               94.0   

       history_of_fever  cough  sore_throat  ...  chronic_hematologic_disease  \
9960                  0      0            0  ...                          0.0   
3880                  0      0            0  ...                          0.0   
10954                 0      1            0  ...                          0.0   
928                   0      0            0  ...                          0.0   
3653                  1      1            1  ...                          0.0   
...                 ...    ...          ...  ...                          ...   
10682                 0      0            0  ...                          0.0   
4975                  0      0            0  ...                          0.0   
10794                 0      0            0  ...                          0.0   
2153                  0      0            0  ...                          0.0   
4277                  0      0            0  ...                          0.0   

       AIDS_HIV  diabetes_mellitus_type_1  diabetes_mellitus_type_2  \
9960        0.0                       0.0                       0.0   
3880        0.0                       0.0                       0.0   
10954       0.0                       0.0                       0.0   
928         0.0                       0.0                       0.0   
3653        0.0                       0.0                       0.0   
...         ...                       ...                       ...   
10682       0.0                       0.0                       0.0   
4975        0.0                       0.0                       0.0   
10794       0.0                       0.0                       0.0   
2153        0.0                       0.0                       0.0   
4277        0.0                       0.0                       0.0   

       rheumatologic_disorder  dementia  tuberculosis  smoking  other_risks  \
9960                      0.0       0.0             0        0            0   
3880                      0.0       0.0             0        0            0   
10954                     0.0       0.0             0        0            0   
928                       0.0       0.0             0        0            0   
3653                      0.0       0.0             0        0            0   
...                       ...       ...           ...      ...          ...   
10682                     0.0       0.0             0 

Así, las nacionalidades que diferían se han convertido en T.C.

### ¿Incoherencias en el género?

In [25]:
data1.groupby('patient ID')['gender K=female E=male'].nunique().value_counts()

gender K=female E=male
1    11122
Name: count, dtype: int64

No existen incoherencias en el género.

Después de estas transformaciones ya hemos tratado los datos que identifican al paciente: nacionalidad, edad y género. Así, cada ID identifica a una sola persona y no existen incoherencias entre los registros de una misma persona.

## Transformación de columnas y tratamiento de nulos

Antes de empezar, las columnas que vamos transformar a valores numéricos son 'nationality', 'gender' y 'PCR_result'.

Y las columnas que contienen datos nulos son las siguientes:

In [26]:
any_null_cols = data1.columns[data1.isnull().any()]
data1[any_null_cols].isnull().sum()

fever_temperature                 468
oxygen_saturation                   4
chronic_kidney_disease              7
obesity                            22
liver_disease                       6
asplenia                           22
chronic_neurological_disorder       2
chronic_hematologic_disease         2
AIDS_HIV                            2
diabetes_mellitus_type_1            3
diabetes_mellitus_type_2            2
rheumatologic_disorder              2
dementia                            2
PCR_result                       1176
dtype: int64

#### PCR_result

Empezaremos por la variable que queremos predecir. Debemos transformarla a datos numéricos. Además, no sería correcto imputar los datos nulos que existen pues esta es la clase en nuestro dataset.

Por tanto, empezamos eliminando los datos nulos.

In [27]:
data1.dropna(subset=['PCR_result'], inplace=True)

Ahora podemos transformar la columna.

In [28]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [29]:
PCR_result_encoder = LabelEncoder()
data1['PCR_result'] = PCR_result_encoder.fit_transform(data1['PCR_result'])

data1['PCR_result'].value_counts()

PCR_result
1    12434
0     1102
Name: count, dtype: int64

La clase tampoco está muy bien balanceada.

#### Nationality

En segundo lugar debemos tratar la columna de nacionalidad, que no contiene nulos. Para ello, antes de nada, debemos explicar el significado de T.C.. Como ya hemos podido ver, el hospital parece turco (por la columna con nombre extravagante) y T.C. puede hacer referencia a "Türkiye Cumhuriyeti", que significa República de Turquía y se suele utilizar en documentos oficiales para identificar a una persona turca.

Además, veamos si las demás nacionalidades está balanceada.

In [30]:
data1['nationality'].value_counts()

nationality
T.C.                              12644
Uzbekistan                           66
United Kingdom (Great Britain)       65
Azerbaijan                           65
Turkmenistan                         60
                                  ...  
El Salvador                           1
Thailand                              1
Rwanda                                1
Somalia                               1
Singapore                             1
Name: count, Length: 97, dtype: int64

Está claro que no. La mayoría de pacientes son turcos y hay hasta 97 países diferentes. De cara a transformar esta columna con un One Hot Encoder, debemos reducir la cantidad de países, ya que la mayoría están poco representados.

Viendo que algunos países tienen más representación en el dataset, podemos dejarlos. Estableceremos el umbral en 60 personas. Es decir, aquellos países con +60 personas en el dataset se mantienen, y los otros los agruparemos en la categoría "Otros".

In [31]:
countries_with_less_than_60_patients = data1['nationality'].value_counts().loc[lambda x: x < 60].index

In [32]:
data1.loc[data1['nationality'].isin(countries_with_less_than_60_patients), 'nationality'] = "Other"

In [33]:
data1['nationality'].value_counts()

nationality
T.C.                              12644
Other                               636
Uzbekistan                           66
Azerbaijan                           65
United Kingdom (Great Britain)       65
Turkmenistan                         60
Name: count, dtype: int64

Así, a pesar de que la clase sigue bastante desbalanceada, ahora no existen países con poca representación. No obstante, tendremos esto en cuenta a la hora de entrenar el modelo.

Ahora podemos pasar a la transformación de la columna de la nacionalidad a datos numéricos: 

In [34]:
nationality_encoder = OneHotEncoder(sparse_output=False)

encoded_nationality = nationality_encoder.fit_transform(data1['nationality'].values.reshape(-1, 1))

encoded_nationality_df = pd.DataFrame(encoded_nationality, columns=nationality_encoder.get_feature_names_out())

data1 = data1.reset_index(drop=True)
data1 = pd.concat([data1['patient ID'], encoded_nationality_df, data1.loc[:, 'age':]], axis=1)
data1

patient ID  x0_Azerbaijan  x0_Other  x0_T.C.  x0_Turkmenistan  \
0        11850006            0.0       0.0      1.0              0.0   
1         4373630            0.0       0.0      1.0              0.0   
2        12172023            0.0       0.0      1.0              0.0   
3        72856168            0.0       0.0      1.0              0.0   
4        72200559            0.0       0.0      1.0              0.0   
...           ...            ...       ...      ...              ...   
13531    98375559            0.0       0.0      1.0              0.0   
13532    35313728            0.0       0.0      1.0              0.0   
13533    52600417            0.0       0.0      1.0              0.0   
13534    72407519            0.0       0.0      1.0              0.0   
13535    83859535            0.0       0.0      1.0              0.0   

       x0_United Kingdom (Great Britain)  x0_Uzbekistan  age  \
0                                    0.0            0.0   47   
1                                    0.0            0.0   23   
2                                    0.0            0.0   26   
3                                    0.0            0.0   60   
4                                    0.0            0.0   72   
...                                  ...            ...  ...   
13531                                0.0            0.0   35   
13532                                0.0            0.0   42   
13533                                0.0            0.0   35   
13534                                0.0            0.0   52   
13535                                0.0            0.0   34   

      gender K=female E=male date_of_first_symptoms  ...  \
0                          K             2022-01-01  ...   
1                          K             2022-01-01  ...   
2                          K             2022-01-01  ...   
3                          K             2022-01-01  ...   
4                          K             2021-12-22  ...   
...                      ...                    ...  ...   
13531                      K             2021-11-01  ...   
13532                      E             2022-03-08  ...   
13533                      K             2021-08-25  ...   
13534                      K             2022-02-04  ...   
13535                      K             2021-10-12  ...   

       chronic_hematologic_disease  AIDS_HIV  diabetes_mellitus_type_1  \
0                              0.0       0.0                       0.0   
1                              0.0       0.0                       0.0   
2                              0.0       0.0                       0.0   
3                              0.0       0.0                       0.0   
4                              0.0       0.0                       0.0   
...                            ...       ...                       ...   
13531                          0.0       0.0                       0.0   
13532                          0.0       0.0                       0.0   
13533                          0.0       0.0                       0.0   
13534                          0.0       0.0                       0.0   
13535                          0.0       0.0                       0.0   

       diabetes_mellitus_type_2  rheumatologic_disorder  dementia  \
0                           0.0                     0.0       0.0   
1                           0.0                     0.0       0.0   
2                           0.0                     0.0       0.0   
3                           0.0                     0.0       0.0   
4                           0.0                     0.0       0.0   
...                         ...                     ...       ...   
13531                       0.0                     0.0       0.0   
13532                       0.0                     0.0       0.0   
13533                       0.0                     0.0       0.0   
13534                       0.0                     0.0       0.0   
1353

#### Género

El género, además de no tener nulos, solo tiene dos posibles valores:

In [35]:
data1['gender K=female E=male'].value_counts()

gender K=female E=male
K    7626
E    5910
Name: count, dtype: int64

Además, se informa en el propio header del significado de cada uno. Con un simple labelEncoder lo transformamos a columna numérica.

In [37]:
gender_encoder = LabelEncoder()
data1['gender K=female E=male'] = gender_encoder.fit_transform(data1['gender K=female E=male'])

In [38]:
data1

patient ID  x0_Azerbaijan  x0_Other  x0_T.C.  x0_Turkmenistan  \
0        11850006            0.0       0.0      1.0              0.0   
1         4373630            0.0       0.0      1.0              0.0   
2        12172023            0.0       0.0      1.0              0.0   
3        72856168            0.0       0.0      1.0              0.0   
4        72200559            0.0       0.0      1.0              0.0   
...           ...            ...       ...      ...              ...   
13531    98375559            0.0       0.0      1.0              0.0   
13532    35313728            0.0       0.0      1.0              0.0   
13533    52600417            0.0       0.0      1.0              0.0   
13534    72407519            0.0       0.0      1.0              0.0   
13535    83859535            0.0       0.0      1.0              0.0   

       x0_United Kingdom (Great Britain)  x0_Uzbekistan  age  \
0                                    0.0            0.0   47   
1                                    0.0            0.0   23   
2                                    0.0            0.0   26   
3                                    0.0            0.0   60   
4                                    0.0            0.0   72   
...                                  ...            ...  ...   
13531                                0.0            0.0   35   
13532                                0.0            0.0   42   
13533                                0.0            0.0   35   
13534                                0.0            0.0   52   
13535                                0.0            0.0   34   

       gender K=female E=male date_of_first_symptoms  ...  \
0                           1             2022-01-01  ...   
1                           1             2022-01-01  ...   
2                           1             2022-01-01  ...   
3                           1             2022-01-01  ...   
4                           1             2021-12-22  ...   
...                       ...                    ...  ...   
13531                       1             2021-11-01  ...   
13532                       0             2022-03-08  ...   
13533                       1             2021-08-25  ...   
13534                       1             2022-02-04  ...   
13535                       1             2021-10-12  ...   

       chronic_hematologic_disease  AIDS_HIV  diabetes_mellitus_type_1  \
0                              0.0       0.0                       0.0   
1                              0.0       0.0                       0.0   
2                              0.0       0.0                       0.0   
3                              0.0       0.0                       0.0   
4                              0.0       0.0                       0.0   
...                            ...       ...                       ...   
13531                          0.0       0.0                       0.0   
13532                          0.0       0.0                       0.0   
13533                          0.0       0.0                       0.0   
13534                          0.0       0.0                       0.0   
13535                          0.0       0.0                       0.0   

       diabetes_mellitus_type_2  rheumatologic_disorder  dementia  \
0                           0.0                     0.0       0.0   
1                           0.0                     0.0       0.0   
2                           0.0                     0.0       0.0   
3                           0.0                     0.0       0.0   
4                           0.0                     0.0       0.0   
...                         ...                     ...       ...   
13531                       0.0                     0.0       0.0   
13532                       0.0                     0.0       0.0   
13533                       0.0                     0.0       0.0   
13534                       0.0                     0.0      

Ahora que ya hemos tratado las 3 columnas que había que transformar, pasemos al tratamiento de nulos.

## Tratamiento de nulos

In [40]:
any_null_cols = data1.columns[data1.isnull().any()]
data1[any_null_cols].isna().sum()

fever_temperature                452
oxygen_saturation                  4
chronic_kidney_disease             4
obesity                           19
liver_disease                      5
asplenia                          19
chronic_neurological_disorder      1
chronic_hematologic_disease        1
AIDS_HIV                           1
diabetes_mellitus_type_1           2
diabetes_mellitus_type_2           1
rheumatologic_disorder             1
dementia                           1
dtype: int64

Observando una a una la distribución de cada columna que contiene al menos un nulo, vemos que en todas menos 'fever_temperature' y 'oxygen_saturation', están muy desbalanceadas, ya que muy poca gente en comparación con el total padece estas enfermedades. Por tanto, las trataremos de diferente manera.

#### Enfermedades (chronic_kidney_disease, obesity, ..., dementia)

Podríamos plantearnos hacer un modelo de clasificación para cada variable binaria. No obstante, pensamos que no merece la pena por dos razones. En primer lugar, estas variables están muy desbalanceadas, y los modelos de clasificación estarían completamente sesgados. Además, la cantidad de valores nulos es muy pequeña.

Por tanto, como el valor dominante en todas estas columnas es 0 (no tiene esa enfermedad), imputaremos con un 0 los nulos en estas columnas.

In [60]:
diseases_to_impute = any_null_cols[(any_null_cols != 'fever_temperature') & (any_null_cols != 'oxygen_saturation')]

In [64]:
data1[diseases_to_impute] = data1[diseases_to_impute].fillna(0)

In [66]:
any_null_cols = data1.columns[data1.isnull().any()]
data1[any_null_cols].isna().sum()

fever_temperature    452
oxygen_saturation      4
dtype: int64

Así, solo nos quedan dos columnas por tratar sus valores nulos.

#### Fever_temperature

Para imputar los valores de 'fever_temperature', intentaremos crear un modelo de regresión con todas las demás variables numéricas, y según su R^2 lo utilizaremos o no.

In [94]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR

fever_temperature_data = data1.dropna(subset=['fever_temperature'])
X = fever_temperature_data.drop(['patient ID', 'fever_temperature', 'date_of_first_symptoms', 'oxygen_saturation'], axis=1)
y = fever_temperature_data['fever_temperature']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Linear Regression
fever_temperature_LR = LinearRegression()
fever_temperature_LR.fit(X_train, y_train)

predicted_fever_temperature_LR = fever_temperature_LR.predict(X_test)

# SVR
fever_temperature_SVR = SVR()
fever_temperature_SVR.fit(X_train, y_train)

predicted_fever_temperature_SVR = fever_temperature_SVR.predict(X_test)

In [95]:
from sklearn.metrics import r2_score

print("El R2 score para el modelo de Regresión Lineal es", r2_score(y_test, predicted_fever_temperature_LR))
print("El R2 score para el modelo de SVR es", r2_score(y_test, predicted_fever_temperature_SVR))

El R2 score para el modelo de Regresión Lineal es -0.003926632515683037
El R2 score para el modelo de SVR es -0.09167568194859


Como vemos, los modelos creados tienen un R2 incluso negativo, es decir, no sirven de absolutamente nada. De hecho, son peores que la imputación con la media. Por tanto, vamos a realizar una imputación con la media.

Antes de imputar nos aseguramos de que no haya ningún dato nulo escondido como un dato normal (por ejemplo, un -1), y para ello veremos el rango en el que se mueve, que deberían de ser valores cercanos a 37.

In [114]:
data1['fever_temperature'].value_counts().sort_index()

fever_temperature
35.000000       2
35.400000       1
35.500000       1
35.700000       1
35.800000       7
35.900000      11
36.000000      20
36.100000      22
36.100000      13
36.200000      36
36.300000     390
36.400000     420
36.500000    1164
36.600000     154
36.600000     658
36.700000     977
36.800000     824
36.900000     517
36.900000     117
37.000000     724
37.100000      72
37.100000     403
37.200000     626
37.300000     475
37.344046     452
37.400000     300
37.400000     130
37.500000     364
37.600000      23
37.600000     240
37.700000     225
37.800000     332
37.900000     392
37.900000      16
38.000000     617
38.100000      38
38.100000     444
38.200000     559
38.300000     328
38.400000     126
38.400000     109
38.500000     105
38.600000      67
38.700000      31
38.800000      82
38.900000     137
38.900000       5
39.000000     302
39.100000     252
39.200000     154
39.300000      49
39.400000      12
39.400000       6
39.500000       2
39.600000 

Los datos se mueven en el rango esperado y no hay datos anómalos, por tanto procedemos a imputar.

In [107]:
data1['fever_temperature'] = data1['fever_temperature'].fillna(data1['fever_temperature'].mean())

#### Oxygen Saturation

In [108]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR

oxygen_saturation = data1.dropna(subset=['oxygen_saturation'])
X = oxygen_saturation.drop(['patient ID', 'fever_temperature', 'date_of_first_symptoms', 'oxygen_saturation'], axis=1)
y = oxygen_saturation['oxygen_saturation']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Linear Regression
oxygen_saturation_LR = LinearRegression()
oxygen_saturation_LR.fit(X_train, y_train)

predicted_oxygen_saturation_LR = oxygen_saturation_LR.predict(X_test)

# SVR
oxygen_saturation_SVR = SVR()
oxygen_saturation_SVR.fit(X_train, y_train)

predicted_oxygen_saturation_SVR = oxygen_saturation_SVR.predict(X_test)

In [109]:
print("El R2 score para el modelo de Regresión Lineal es", r2_score(y_test, predicted_oxygen_saturation_LR))
print("El R2 score para el modelo de SVR es", r2_score(y_test, predicted_oxygen_saturation_SVR))

El R2 score para el modelo de Regresión Lineal es -0.008764786272094316
El R2 score para el modelo de SVR es -0.048371842355942496


De nuevo negativos, por tanto los modelos no son capaces de explicar la variabilidad y debemos buscar otro método. De nuevo, utilizaremos la media.

Antes, comprobaremos si hay algún valor no nulo que realmente sí lo sea, viendo de nuevo el rango. El nivel de saturación de oxígeno normal en una persona sana es entre el 95% y el 100%, por tanto, la mayoría de los valores deberían moverse entre esos números.

In [113]:
data1['oxygen_saturation'].value_counts().sort_index()

oxygen_saturation
-1.0        15
 0.0        59
 69.0        8
 70.0       11
 71.0        4
 76.0        3
 77.0       17
 78.0       41
 79.0      102
 80.0      137
 81.0       72
 82.0       21
 83.0       13
 84.0       30
 85.0      113
 86.0      181
 87.0      246
 88.0      515
 89.0      570
 90.0      380
 91.0      192
 92.0      144
 93.0      112
 94.0      835
 95.0     3038
 96.0     3345
 97.0     1675
 98.0     1122
 99.0      480
 100.0      51
Name: count, dtype: int64

En primer lugar vemos que hay observaciones con un nivel de -1 y 0. Ambos valores son inválidos, y los consideraremos nulos.

In [116]:
import numpy as np

In [119]:
data1.loc[data1['oxygen_saturation'].isin([-1, 0]), 'oxygen_saturation'] = np.nan

In [120]:
data1['oxygen_saturation'].value_counts().sort_index()

oxygen_saturation
69.0        8
70.0       11
71.0        4
76.0        3
77.0       17
78.0       41
79.0      102
80.0      137
81.0       72
82.0       21
83.0       13
84.0       30
85.0      113
86.0      181
87.0      246
88.0      515
89.0      570
90.0      380
91.0      192
92.0      144
93.0      112
94.0      835
95.0     3038
96.0     3345
97.0     1675
98.0     1122
99.0      480
100.0      51
Name: count, dtype: int64

Ahora sí, observamos los valores reales de la distribución. La mayoría se mueven entre 94 y 99%. No obstante, hay valores muy bajos cercanos a 70, lo cual puede tener que ver con alguna enfermedad grave.

In [121]:
data1['oxygen_saturation'].isna().sum()

78

Hemos pasado de 4 a 78 nulos. Como se hizo antes, imputaremos con la media pues los modelos de regresión han resultado poco eficaces.

In [122]:
data1['oxygen_saturation'] = data1['oxygen_saturation'].fillna(data1['oxygen_saturation'].mean())

Así, ya no existen nulos en los datos.

Antes de terminar con los nulos, hemos visto que existen valores con -1. Comprobemos que no haya más en todo el dataset.

In [132]:
data1[data1 == -1].count()

patient ID                           0
x0_Azerbaijan                        0
x0_Other                             0
x0_T.C.                              0
x0_Turkmenistan                      0
x0_United Kingdom (Great Britain)    0
x0_Uzbekistan                        0
age                                  5
gender K=female E=male               0
date_of_first_symptoms               0
fever_temperature                    0
oxygen_saturation                    0
history_of_fever                     0
cough                                0
sore_throat                          0
runny_nose                           0
wheezing                             0
shortness_of_breath                  0
lower_chest_wall_indrawing           0
chest_pain                           0
conjunctivitis                       0
lymphadenopathy                      0
headache                             0
loss_of_smell                        0
loss_of_taste                        0
fatigue_malaise          

Existen 5 valores de -1 en la columna edad. Vamos a imputarlos con la media. 

In [133]:
data1.loc[data1['age'].isin([-1]), 'age'] = np.nan

In [134]:
data1['age'] = data1['age'].fillna(data1['age'].mean())

In [135]:
data1['age'].value_counts().sort_index()

age
0.0       9
1.0      27
2.0      46
3.0      42
4.0      55
         ..
95.0      4
96.0      4
97.0      2
98.0      1
100.0     1
Name: count, Length: 101, dtype: int64

Así, presuntamente todos los valores del dataset son válidos.

In [148]:
data1.groupby('patient ID')['date_of_first_symptoms'].nunique()

patient ID
568         2
661         1
3982        1
9490        1
11368       1
           ..
99939860    1
99952878    1
99960888    1
99972402    1
99980488    1
Name: date_of_first_symptoms, Length: 10326, dtype: int64

In [157]:
pd.set_option('display.max_columns', None)

In [158]:
data1[data1.duplicated(subset=['patient ID', 'date_of_first_symptoms'], keep=False)].sort_values(by=['patient ID', 'date_of_first_symptoms'])

patient ID  x0_Azerbaijan  x0_Other  x0_T.C.  x0_Turkmenistan  \
10596    17630916            0.0       0.0      1.0              0.0   
13499    17630916            0.0       0.0      1.0              0.0   
1027     24545730            0.0       0.0      1.0              0.0   
13227    24545730            0.0       0.0      1.0              0.0   
6361     27701770            0.0       1.0      0.0              0.0   
10440    27701770            0.0       1.0      0.0              0.0   
790      29967900            0.0       0.0      1.0              0.0   
4731     29967900            0.0       0.0      1.0              0.0   
563      70754545            0.0       0.0      1.0              0.0   
1660     70754545            0.0       0.0      1.0              0.0   
502      80743290            0.0       0.0      1.0              0.0   
8427     80743290            0.0       0.0      1.0              0.0   
9115     80743290            0.0       0.0      1.0              0.0   

       x0_United Kingdom (Great Britain)  x0_Uzbekistan   age  \
10596                                0.0            0.0  42.0   
13499                                0.0            0.0  42.0   
1027                                 0.0            0.0  10.0   
13227                                0.0            0.0  10.0   
6361                                 0.0            0.0  54.0   
10440                                0.0            0.0  54.0   
790                                  0.0            0.0  42.0   
4731                                 0.0            0.0  42.0   
563                                  0.0            0.0  24.0   
1660                                 0.0            0.0  24.0   
502                                  0.0            0.0  21.0   
8427                                 0.0            0.0  21.0   
9115                                 0.0            0.0  21.0   

       gender K=female E=male date_of_first_symptoms  fever_temperature  \
10596                       1             2021-08-24               36.9   
13499                       1             2021-08-24               37.9   
1027                        1             2021-12-30               37.7   
13227                       1             2021-12-30               37.2   
6361                        1             2022-02-14               38.0   
10440                       1             2022-02-14               37.9   
790                         1             2021-08-02               38.1   
4731                        1             2021-08-02               38.0   
563                         1             2022-02-02               38.2   
1660                        1             2022-02-02               37.1   
502                         0             2021-12-15               36.5   
8427                        0             2021-12-15               36.4   
9115                        0             2021-12-15               36.5   

       oxygen_saturation  history_of_fever  cough  sore_throat  runny_nose  \
10596               96.0                 0      0            0           0   
13499               95.0                 1      0            0           0   
1027                96.0                 1      0            0           0   
13227               98.0                 0      0            0           0   
6361                98.0                 0      0            0           0   
10440               95.0                 0      0            0           0   
790                 97.0                 1      1            1           1   
4731                93.0                 0      0            0           0   
563                 88.0                 0      0            0           0   
1660                95.0                 0      0            0           0   
502                 95.0                 0      0            0           0   
8427                96.0                 0      0            0           0   
9115                97

#### ¿FINAL DATASET 1?

 - Mirar acumulaciones de visitas.